crates/subspace-networking/src/reserved_peers.rs

```diff
@@ -0,0 +1,254 @@
+#[derive(Debug, Clone)]
+pub struct Event;
+
+impl Behaviour {
+    /// Creates a new `Behaviour` with a predefined set of reserved peers.
+    pub fn new(config: Config) -> Self {
+        debug!(
+            reserved_peers=?config.reserved_peers,
+            "Reserved peers protocol initialization...."
+        );
+
+        let peer_addresses = convert_multiaddresses(config.reserved_peers);
+
+        let reserved_peers_state = peer_addresses
+            .iter()
+            .cloned()
+            .map(|(peer_id, address)| {
+                (
+                    peer_id,
+                    ReservedPeerState {
+                        peer_id,
+                        address,
+                        connection_status: ConnectionStatus::NotConnected {
+                            scheduled_at: schedule_connection(),
+                        },
+                    },
+                )
+            })
+            .collect();
+
+        Self {
+            protocol_name: config.protocol_name,
+            reserved_peers_state,
+        }
+    }
+
+    /// Create a connection handler for the reserved peers protocol.
+    #[inline]
+    fn new_reserved_peers_handler(&self, peer_id: &PeerId) -> Handler {
+        Handler::new(
+            self.protocol_name,
+            self.reserved_peers_state.contains_key(peer_id),
+        )
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ConnectionHandler = Handler;
+    type OutEvent = Event;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _: ConnectionId,
+        peer_id: PeerId,
+        _: &Multiaddr,
+        _: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        Ok(self.new_reserved_peers_handler(&peer_id))
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _: ConnectionId,
+        peer_id: PeerId,
+        _: &Multiaddr,
+        _: Endpoint,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        Ok(self.new_reserved_peers_handler(&peer_id))
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        match event {
+            FromSwarm::ConnectionEstablished(ConnectionEstablished { peer_id, .. }) => {
+                if let Some(state) = self.reserved_peers_state.get_mut(&peer_id) {
+                    state.connection_status = ConnectionStatus::Connected;
+
+                    debug!(peer_id=%state.peer_id, "Reserved peer connected.");
+                }
+            }
+            FromSwarm::ConnectionClosed(ConnectionClosed {
+                peer_id,
+                remaining_established,
+                ..
+            }) => {
+                if let Some(state) = self.reserved_peers_state.get_mut(&peer_id) {
+                    if remaining_established > 0 {
+                        state.connection_status = ConnectionStatus::Connected;
+                    } else {
```

Comment:
Does it need to be updated though? If there is one left, it means it is already connected.

```suggestion
                    if remaining_established == 0 {
```

Also is it possible that this connection has closed, but there is another one in progress and you should wait for it rather than creating yet another one that could fail due to exceeding connection limit per peer?

Reply:
We will wait for it because we don't immediately connect but schedule a connection for the future (1 sec delay). Pending connections could can that. We could increase the delay to give potential pending connections more time. Also, I added `remaining_established > 0` condition here because we could have an additional established connection and it would be enough to consider peer connected.

Reply:
What I meant is that if we have more than one connection, then clearly `state.connection_status == ConnectionStatus::Connected` already. Setting it to the same value is redundant.

1 second delay will not prevent extra connections from being established, but I guess it is fine. The only thing that isn't fine is that `FromSwarm::DialFailure` below will set connection status to `NotConnected` even if this new dial fails, but other connection was established successfully. I think you need to check in there that current state is `PendingConnection` and otherwise leave it what it is, just like you're checking in `fn poll` implementation.